### ISSUE TITLE
Can Renarde OIDC automatically redirect an @Authenticated method to Login upon an HTTP 401?

### ISSUE THREAD

**Issue by preslavrachev** [quarkiverse/quarkus-renarde#30]
I searched throughout the entire documentation, but could not find an easy option to do so. As a result, hitting an `@Authenticated` method simply returns a default browser error page (HTTP 401). In my opinion, the graceful behavior in this case should be redirecting to the login page (specified via a property).

I temporarily solved this by extending and overriding parts of the `JWTAuthMechanism` class:

```java
@Alternative
@Priority(1)
@ApplicationScoped
public class JWTAuth extends JWTAuthMechanism {

   @Override
   public Uni<ChallengeData> getChallenge(RoutingContext context) {
      return Uni.createFrom().item(new ChallengeData(302, "Location", "/login"));
   }
}
```

I suppose this is something that could make it to a future release of Renarde.

**Comment by FroMage** [comment]
I did notice weird things related to combining JWT/OIDC/WebAuthn in this regard, but I thought they were walking on eachother's toes.
I should look more into it.

**Comment by sberyozkin** [comment]
@preslavrachev @FroMage if 3 authentication methods are available, then, when a new request without any credentials arrives, there is no information available which one to prioritize. They co-exist together fine if the credentials are supplied in the request, but if nothing is available, then Quarkus Security will just pick the first one it finds.

**Comment by sberyozkin** [comment]
But in this case a custom challenge is suggested, so I think the proposed solution is fine, it just customizes the default behavior since responding 302 to the JWT bearer authentication is not a typical case, not all the clients will be using the user agents understanding 302.

**Comment by FroMage** [comment]
Related to https://github.com/quarkusio/quarkus/issues/28247

**Comment by tmulle** [comment]
Is this still in the works? I just hit this using the renarde security JWT only.

My endpoint is defined with:
```
@RolesAllowed({"webadmin", "development", "support"})
    @GET
    @Path("/admin/versions")
    @Produces(MediaType.TEXT_HTML)
```

and when I try to use the browser to `http://localhost:8080/admin/versions` without being logged in, I see a blank screen and the response is a 401 if I look at the network response in the dev tools.

I would also like to have a redirect back to login page. This is what our old (play/scala) app did. And I need it to work the same for our conversion to be complete.

I have this set according to the docs (if it helps):
```
# Must be set to get proper error reporting
quarkus.http.auth.proactive=false
```

Thanks..

**Comment by FroMage** [comment]
What version are you using? It should work in 1.0.5 or 3.0.0

**Comment by tmulle** [comment]
I’m using 1.0.5 with Quarkus 2.16.4.Final

**Comment by tmulle** [comment]
@FroMage Yes, we are using our own custom forms to login and my endpoints are defined as.
We are using username/password against a database, that's the way the legacy app was developed and that's what I had to maintain for now.

```
@GET
    @Path("/login")
    @Produces(MediaType.TEXT_HTML)
    public Response login() {
        return ok(Templates.login(new LoginFormData()));
    }


    @POST
    @Path("/login")
    @Consumes(MediaType.APPLICATION_FORM_URLENCODED)
    @Produces(MediaType.TEXT_HTML)
    public Response doLogin(LoginFormData loginForm) { ...
```

**Comment by FroMage** [comment]
OK, so, a few tips:

- you can drop the `@GET` annotation, it's implicit unless you have others like `@POST`.
- you can drop the `@Produces(HTML)` annotations, they're not useful when returning templates or redirecting
- you can also drop the `Consumes(MediaType.APPLICATION_FORM_URLENCODED)` because I assume `LoginFormData` contains `@RestForm` or `@FormParam` members
- you can return `TemplateInstance` instead of `Response`, it's nicer and shorter

Next, you can try setting the `@LoginPage` annotation on your `login` (get) method, this should get you your redirect. Let me know if it works and that's all you need?

So, to recap:

```java
    @LoginPage
    @Path("/login")
    public TemplateInstance login() {
        return Templates.login(new LoginFormData());
    }


    @POST
    @Path("/login")
    public Response doLogin(LoginFormData loginForm) { ...
```

**Comment by tmulle** [comment]
Thanks for the tips!

Yeah, I knew about the implicit @GET, etc. I just like to put them there to be consistent so others besides myself in the code will be able to know what the function is doing.

I go back and forth on whether I should return a `Response` or `TemplateInstance` the legacy app I'm converting from Scala/Play always return a `Result` object which I think is like the `Response` object, so I started doing that.

The also would do things, like return `badRequest(<JSONDATA>)` or sometimes `badRequest(<HTML>)`  or `ok(<JSONDATA>)` back to their web front end which usually called the backed via Ajax in javascript. It was quite a mess the way they kept mixing return types.

Using the @LoginPage worked! When I no try to directly hit in the browser an `Authenticated` or `RolesAllowed` endpoint that returns HTML it redirects to the login page, very cool!

I didn't change my `RenardeUser` like you mentioned, is that still necessary? It seems to work without changing my User class.

**Comment by FroMage** [comment]
> Using the https://github.com/loginpage worked! When I now try to directly hit in the browser an Authenticated or RolesAllowed endpoint that returns HTML it redirects to the login page, very cool!

OK great!

> I didn't change my RenardeUser like you mentioned, is that still necessary? It seems to work without changing my User class.

You don't need to. If you do, Renarde will handle the login page itself for you, but you already have endpoints for it, so you don't need to.

It's a bit hard to customise ATM, so it's still experimental :)

**Comment by FroMage** [comment]
I'm going to close this from old age, feel free to reopen if you still have the issue.
